src/charm.py

```diff
@@ -263,6 +253,38 @@ def ready_to_configure(self) -> bool:
 
         return True
 
+    def _check_and_update_certificate(self) -> bool:
+        """Check if the certificate or private key needs an update and perform the update.
+
+        This method retrieves the currently assigned certificate and private key associated with
+        the charm's TLS relation. It checks whether the certificate or private key has changed
+        or needs to be updated. If an update is necessary, the new certificate or private key is
+        stored.
+
+        Returns:
+            bool: True if either the certificate or the private key was updated, False otherwise.
+        """
+        provider_certificate, private_key = self._certificates.get_assigned_certificate(
+            certificate_request=self._get_certificate_request()
+        )
+        if not provider_certificate or not private_key:
+            logger.debug("Certificate or private key is not available")
+            return False
+        if certificate_update_required := self._is_certificate_update_required(
+            provider_certificate.certificate
+        ):
+            self._store_certificate(certificate=provider_certificate.certificate)
+        if private_key_update_required := self._is_private_key_update_required(private_key):
+            self._store_private_key(private_key=private_key)
+        return certificate_update_required or private_key_update_required
+
+    @staticmethod
+    def _get_certificate_request() -> CertificateRequest:
+        return CertificateRequest(
+            common_name=CERTIFICATE_COMMON_NAME,
+            sans_dns=frozenset([CERTIFICATE_COMMON_NAME]),
```

Comment:
Looking at this and the hardcoded value, I think this might cause issues if the model if the model is not named `sdcore` and the app not named `smf`.

Reply:
Yes we use hardocded values here but this was not implemented in this PR, the same values were used with V3. We probably want to be smarter with the content of those certs but this should likely be done independently of this change here.